test-snapshots/src/snapshots-input/Snapshot.Literals.String.purs

```diff
@@ -0,0 +1,86 @@
+module Snapshot.Literals.String where
+
+block1 :: String
+block1 = """block"""
+
+block2 :: String
+block2 = """
+foo
+bar
+"""
+
+block3 :: String
+block3 = """foo
+bar"""
+
+block4 :: String
+block4 = """foo
+bar
+baz"""
+
+block5 :: String
+block5 = """foo
+bar
+baz
+"""
+
+---
+
+escape1 :: String
+escape1 = "\t"
+
+escape2 :: String
+escape2 = "\n"
+
+escape3 :: String
+escape3 = "\r"
+
+escape4 :: String
+escape4 = "\r"
+
+escape5 :: String
+escape5 = "\r\n"
+
+escape6 :: String
+escape6 = "\\"
+
+escape7 :: String
+escape7 = "\""
+
+escape8 :: String
+escape8 = "\'"
+
+---
+
+unicode1 :: String
+unicode1 = "\x0"
+
+-- unicode2 :: String
+-- unicode2 = "\x10ffff"
+--
+-- The above definition causes the build to fail.
+--
+-- [43 of 43] Building Snapshot.Literals.String
+-- file:///home/u/purescm/output/Effect.Aff/foreign.js:530
+--                 throw util.fromLeft(step);
+--                 ^
```

Comment:
If you change [this line](https://github.com/purescm/purescm/blob/master/test/Main.purs#L136) to
```purs
hasMain <- either (unsafeCrashWith <<< append "canRunMain: parse failure - ") pure $
  canRunMain originalFileSourceCode
```
what does it output?

Reply:
It's hitting the ParseFailed case.

Reply:
Let's change the `Test.Main` to not have that `unsafeCrash` so we can always know what's the relevant error.

Reply:
I should have printed the full error when I added this initially...
```
Error: canRunMain: parse failure

Could not parse file: 
position: { column: 20, line: 2 }
error: Character escape out of range \10ffff

Reply:
Pretty sure this is a bug in `cst-parser` because it's not using [`enums@v6.01`](https://github.com/purescript/purescript-enums/releases). The `enums` fix was [released on 2022-08-16](https://github.com/purescript/purescript-enums/releases) but `cst-parser` is using a [package set that was released on 2022-05-07](https://github.com/natefaubion/purescript-language-cst-parser/blob/main/packages.dhall#L2)

Reply:
I could be wrong, but that would be my first guess.

Reply:
The CST Parser uses `Char.fromCharCode`, so if it's a bug, it's a bug there.
https://github.com/natefaubion/purescript-language-cst-parser/blob/13a2437571b60c2c861bbd5b0d07850b4cc82e94/src/PureScript/CST/Lexer.purs#L527

Reply:
> The enums fix was [released on 2022-08-16](https://github.com/purescript/purescript-enums/releases) but cst-parser is using a [package set that was released on 2022-05-07](https://github.com/natefaubion/purescript-language-cst-parser/blob/main/packages.dhall#L2)

It wouldn't matter what cst-parser is using internally. What matters is the version you are using in this project.

Reply:
I think `Char.fromCharCode` is buggy: https://try.purescript.org/?code=FAWw9gJgrgNgpgAgLIEMCWA7BB3AFnAJzmGDRAAcwCAXBABSJigmNIqtoBEVqUA6VAE8ARogAUQ0WL58AlABoEAMwJgQkuIpVqAUlADO1RSBQi4stpRoIAokqVwAxrTF2Hzi2Su03T6nwBhGBR9fUCwDH0weAQxGDAAc092awAVAkE6KCIAZUcCNHJaEIR0zOy4PIKiyw4Ebl5A3BQCBBKA5oJa6wb%2BAEkMYv0EAeoSE0wEAC4p23s-BABVDDQxiawAXlKMrNz8wv8iDBZWjYAeM%2B3yver-bFXcAIiomIgwYAQEeISEABI28jkODHBAAIgAjAAGABisOhbWGLEcZBQMFmnwxmIxoL%2BCH0uDA2Fxoz42hAOWoBQwCQAgsMSfgAB4oJEomBgqH2eygj5fRK4lCA4EQMEdFp8ahgMUEJ4sWKS8jTWbS2SzHH-fGE3F6Qy46USqWdWWIBW874CoUg0GcNDzI6OODorHOl3O9V4glE-7gn0AFh94IQAFoEAA2ACs4YAzOHeUA

```
10FFFF as decimal:             (Just 1114111)
Char.toCharCode (top :: Char): (Just 65535)
Difference:                          1048576
```

Reply:
So. `Char.fromCharCode` is correct per [Prim docs](https://pursuit.purescript.org/builtins/docs/Prim#t:Char):
> A single character (UTF-16 code unit). The JavaScript representation is a normal String, which is guaranteed to contain one code unit. This means that astral plane characters (i.e. those with code point values greater than 0xFFFF) cannot be represented as Char values.

The issue here is that we're parsing something beyond that. In actuality, we need to use CodePoints to accurately parse this. Per [CodePoint docs](https://pursuit.purescript.org/packages/purescript-strings/6.0.1/docs/Data.String.CodePoints#t:CodePoint):
> CodePoint is an Int bounded between 0 and 0x10FFFF, corresponding to Unicode code points.

The fix is to use `(toEnum :: Int -> Maybe CodePoint)`. Proof via Try PS: https://try.purescript.org/?code=FAWw9gJgrgNgpgAgLIEMCWA7BB3AFnAJzmGDRAAcwCAXBABSJigmNIqtoBEVqUA6VAE8ARogAUQ0WL58AlABoEAMwJgQkuIpVqAUlADO1RSBQi4stpRoIAokqVwAxrTF2Hzi2Su03T6nwBhGBR9fUCwDH0weAQxGDAAc092awAVAkE6KCIAZUcCNHJaEIR0zOy4PIKiyw4Ebl5A3BQCBBKA5oJa6wb%2BAEkMYv0EAepurh5%2BHOoCjATwljowTGphkpyAunH6yb4bDCgQWOowfcOLUHQsAC5r23s-BABVDDQxk0wEAF5SjKzc-KFfxEDAsVpfAA8EN%2B5QB1X82DeuACESiMQgYGACAQ8QSCAAJG1yOQ4KCEAAiACMAAYAGL02ltYYsRxkFAwO7YrncrnkgkIfS4MDYfmjPjaEDTWYJACCwzF%2BAAHigWWyYBSafZ7OSsTjEvyUMTSRAKR0WnwTmaCCiWMcwOQELcEFbZHc%2BYTBcL%2BXpDPyrRawFabYgTuRdbiDUayeTOGgHiDHHBOTyU6mU%2B6BUKRYTKbmACy5ykIAC0CAAbABWCsAZgr4f1hMNJOjT30mDxGzobv5nuzCBMDs7fDbc3gJywYhOZyOhYLucddw0CCHwaWKws2KAA

```
main :: Effect Unit
main = TryPureScript.render =<< TryPureScript.withConsole do
  log $ append "10FFFF as decimal:             " $ show $ Int.fromStringAs Int.hexadecimal "10ffff"
  log $ append "Char.toCharCode (top :: Char): " $ show $ Just $ Char.toCharCode top
  log $ append "Difference:                          " $ show $ 1114111 - 65535
  log $ append "Using SCP: " $ show $ map SCP.singleton (toEnum 1114111 :: Maybe SCP.CodePoint)
```

Reply:
I've read the lexer code a bit. If I understanc correctly the current version of parseEscape which eventually calls parseHexEscape is used by both parseStringEscape (which is called by parseStringPart and parseString) and by parseChar, but only works for the latter. parseStringEscape should use a function that uses CodePoint instead of Char. Right?

Reply:
So the issue here is that the compiler has two different constraints for character escapes:
* When used in a Char literal there is an upper bound of 0xFFFF
* When used in a String literal there is an upper bound of 0x10FFF

And `cst-parser` is using the 0xFFFF upper bound for both.

Reply:
https://github.com/natefaubion/purescript-language-cst-parser/issues/48